### PR TITLE
lis2dh: add tap interrupt

### PR DESCRIPTION
**Pull request by mathiasstorck**
Add interrupt for single tap on ST LIS2DH driver  
Tap interrupt is shared with anym interrupt  
Slope threshold and duration is used to configure any motion and click interrupt

### CODE REVIEW

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:471]: Do a better indentetaion here

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh_trigger.c]: better indentation

**Review by avisconti — CHANGES_REQUESTED** [review]
Looks almost good to me. Pls take a look at indentation remarks.
moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).

**Comment by gwa-storck** [comment]
> Looks almost good to me. Pls take a look at indentation remarks. moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).

Thanks, you are right. Just corrected some more indentations. So they should be squashed when merged

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:53]: argument should be aligned to 'dev', like it was before. Do it globally

**Review by avisconti — CHANGES_REQUESTED** [review]

**Comment by avisconti** [comment]
> > Looks almost good to me. Pls take a look at indentation remarks. moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).
> 
> Thanks, you are right. Just corrected some more indentations. So they should be squashed when merged

How the commits are automatically squashed together?

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (gwa-storck) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:53]: it was a mixup before, some with alignment to the first parameter:
https://github.com/zephyrproject-rtos/zephyr/blob/4129b10b6900133e41829893c36a01527f293548/drivers/sensor/lis2dh/lis2dh_trigger.c#L150
and some with a tab indentation:
https://github.com/zephyrproject-rtos/zephyr/blob/4129b10b6900133e41829893c36a01527f293548/drivers/sensor/lis2dh/lis2dh_trigger.c#L155

Now I changed it globally to the tab indentation. I think it is a matter of taste

**Comment by gwa-storck** [comment]
> > > Looks almost good to me. Pls take a look at indentation remarks. moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).
> > 
> > 
> > Thanks, you are right. Just corrected some more indentations. So they should be squashed when merged
> 
> How the commits are automatically squashed together?

I think they should get squashed by the one who merges. Never published to github before, only to gitlab. But when it works the same the one that merges should do this

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:53]: > Now I changed it globally to the tab indentation. I think it is a matter of taste

AFAIK it's a requirement.

**Comment by avisconti** [comment]
> > > > Looks almost good to me. Pls take a look at indentation remarks. moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).
> > > 
> > > 
> > > Thanks, you are right. Just corrected some more indentations. So they should be squashed when merged
> > 
> > 
> > How the commits are automatically squashed together?
> 
> I think they should get squashed by the one who merges. Never published to github before, only to gitlab. But when it works the same the one that merges should do this

Mmmmh, not sure about this. I guess the merger is expecting the committer and reviewer to prepare a good patchset before merging. Maybe @MaureenHelm may clarify this.

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (gwa-storck) [reply drivers/sensor/lis2dh/lis2dh_trigger.c]: my problem was that I had TAB set to 4 spaces, corrected it now

**Comment by MaureenHelm** [comment]
> > > > > Looks almost good to me. Pls take a look at indentation remarks. moreover, pls squash the two commits in a single one, as it seems to me that the second is just a correction of the first (if I'm not wrong).
> > > > 
> > > > 
> > > > Thanks, you are right. Just corrected some more indentations. So they should be squashed when merged
> > > 
> > > 
> > > How the commits are automatically squashed together?
> > 
> > 
> > I think they should get squashed by the one who merges. Never published to github before, only to gitlab. But when it works the same the one that merges should do this
> 
> Mmmmh, not sure about this. I guess the merger is expecting the committer and reviewer to prepare a good patchset before merging. Maybe @MaureenHelm may clarify this.

The author needs to do the squashing.

**Comment by gwa-storck** [comment]
I think I'm confused ;)
Just rebased the changes to get them into a squashed commit. 
Chose pick for the initial commit and fixup all follow ups.
And now there is the failure in the pipeline:
```
Rebasing (1/6)
Auto-merging drivers/sensor/lis2dh/lis2dh.h
CONFLICT (content): Merge conflict in drivers/sensor/lis2dh/lis2dh.h
Auto-merging drivers/sensor/lis2dh/lis2dh_trigger.c
CONFLICT (content): Merge conflict in drivers/sensor/lis2dh/lis2dh_trigger.c
Rebasing (2/6)
error: could not apply 19ef0967[31](https://github.com/zephyrproject-rtos/zephyr/actions/runs/6168315375/job/16740627170#step:7:32)... lis2dh: add tap interrupt
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 19ef096731... lis2dh: add tap interrupt
Error: Process completed with exit code 1.
```

What do I need to do?

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:262]: Use a separate set routine here. Alternatively, use a single one, but with a different name (anym_set seems only for "any motion"). I would be more for a separation if not too difficult.

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh_trigger.c:209]: ok, so you are using the presence of an handler (set by SENSOR_TRIG_DELTA and/or SENSOR_TRIG_TAP) to start it. Good

**Review by avisconti — CHANGES_REQUESTED** [review]
I'm almost ok with this commit. You also made better original code, thanks!
Please provide your feedback on my new comments and requests. We are almost there.

File: drivers/sensor/lis2dh/lis2dh_trigger.c
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:209]: BTW, have you tested both TAP and ANY MOTION just to see if both works correctly? Please do... (thx!)

Comment (gwa-storck) [reply drivers/sensor/lis2dh/lis2dh_trigger.c:224]: yes, both were tested and are worked as expected

File: drivers/sensor/lis2dh/lis2dh.h
Comment (avisconti) [inline drivers/sensor/lis2dh/lis2dh.h:161]: I think we can remove also this one

**Review by avisconti — CHANGES_REQUESTED** [review]
I think this is last one

File: drivers/sensor/lis2dh/lis2dh.h
Comment (avisconti) [reply drivers/sensor/lis2dh/lis2dh.h:161]: ok, I see

**Review by avisconti — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaxMaeder** [comment]
Has anyone been able to successfully use this? I've modified the lis2dh sample to detect taps, and I can't get it to detect anything, as in my handler never fires.

I'd really appreciate any help getting this working, as there isn't a ton of documentation around this.

Here is my modified code:
```conf
CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
CONFIG_I2C=y
CONFIG_SENSOR=y
CONFIG_LIS2DH=y
CONFIG_LIS2DH_TRIGGER_GLOBAL_THREAD=y
```
```c
#include <stdio.h>
#include <zephyr/kernel.h>
#include <zephyr/device.h>
#include <zephyr/drivers/sensor.h>

static void trigger_handler(const struct device *dev,
                            const struct sensor_trigger *trig)
{
    printf("TRIG\n");
}

int main(void)
{
    const struct device *sensor = DEVICE_DT_GET(DT_ALIAS(accel0));

    if (sensor == NULL || !device_is_ready(sensor))
    {
        printf("Could not get accel0 device\n");
        return -ENODEV;
    }

    struct sensor_trigger trig;
    int rc;

    trig.type = SENSOR_TRIG_TAP;
    rc = sensor_trigger_set(sensor, &trig, trigger_handler);
    if (rc != 0)
    {
        printf("Failed to set trigger: %d\n", rc);
        return rc;
    }

    printf("Waiting for triggers\n");
    while (true)
    {
        k_sleep(K_MSEC(2000));
    }

    return 0;
}
```

**Pull request merged**
